keps/sig-network/2091-cluster-network-policy/README.md

```diff
@@ -0,0 +1,958 @@
+### Shared API Design
+The following structs will be added to the `networking.k8s.io` API group and
+shared between `ClusterNetworkPolicy` and `DefaultNetworkPolicy`:
+
+```golang
+type AppliedTo struct {
+	// required if a PodSelector is specified
+	NamespaceSelector   *metav1.LabelSelector
+	// optional
+	PodSelector         *metav1.LabelSelector
+}
+
+type Namespaces struct {
+	// Self and Selector are mutually-exclusive
+	Self       bool
+	Selector   *metav1.LabelSelector
+}
+```
+
+#### AppliedTo
+The `AppliedTo` field replaces `PodSelector` in NetworkPolicy spec, as means to specify
+the target Pods that this cluster-scoped policy (either `ClusterNetworkPolicy` or
+`DefaultNetworkPolicy`) applies to.
+Since the policy is cluster-scoped, the `NamespaceSelector` field is required.
+An empty `NamespaceSelector` (namespaceSelector: {}) selects all Namespaces in the Cluster.
+
+#### Namespaces
+The `Namespaces` field replaces `NamespaceSelector` in NetworkPolicyPeer, as
+means to specify the Namespaces of ingress/egress peers for cluster-scoped policies.
+For selecting Pods from specific Namespaces, the `Selector` field under `Namespaces`
+works exactly as `NamespaceSelector`. The `Self` field is added to satisfy the
+specific needs for cluster-scoped policies:
+
+__Self:__ An optional field, which evaluates to false by default.
+When `self: true` is set, no selectors can be present concurrently. This is a
+special keyword to indicate that the rule only applies to the Namespace for
+which the ingress/egress rule is currently being evaluated upon. Since the Pods
+selected by the ClusterNetworkPolicy appliedTo could be from multiple Namespaces,
+the scope of ingress/egress rules whose `namespace.self=true` will be the Pod's
+own Namespace for each selected Pod.
+Consider the following example:
+
+- Pods [a1, b1] exist in Namespace x, which has labels `app=a` and `app=b` respectively.
+- Pods [a2, b2] exist in Namespace y, which also has labels `app=a` and `app=b` respectively.
+
+```yaml
+apiVersion: networking.k8s.io/v1alpha1
+kind: DefaultNetworkPolicy
+spec:
+  appliedTo:
+    namespaceSelector: {}
+  ingress:
+    - onlyFrom:
+      - namespaces:
+          self: true
+        podSelector:
+          matchLabels:
+            app: b
+```
+
+The above DefaultNetworkPolicy should be interpreted as: for each Namespace in
+the cluster, all Pods in that Namespace should only allow traffic from Pods in
+the _same Namespace_ who has label app=b. Hence, the policy above allows
+x/b1 -> x/a1 and y/b2 -> y/a2, but denies y/b2 -> x/a1 and x/b1 -> y/a2.
+
+#### IPBlock
+
+The `ClusterNetworkPolicyPeer` and `DefaultNetworkPolicyPeer` both allow the
+ability to set an `IPBlock` as a peer. The usage of this field is similar to
+how it is used in the NetworkPolicyPeer. However, we should also explicitly
+note that the IPBlock set in this field could belong to the cluster locally, or
+could be cluster external. For example, the peer could be set with a subnet
```

Comment:
_The first user story in this KEP_ is "As a cluster admin, I want to explicitly deny traffic from certain source IPs that I know to be bad." If you leave `ipBlock` behavior implementation-defined, you are not solving that user story. This is a crucial feature. We can't just punt on it.

Reply:
+1 here, one of the major problems I've seen with Network Policy is leaving certain behaviors of selectors (like IPBLOCK) as implementation Specific. This results in many caveats across CNI's.... If the behavior is something that can easily be clearly defined I think it should be.

Reply:
> the source IP seen by the CNI implementation is what will be considered

I think it needs to be stronger than this. CNI isn't necessarily Service aware, but NetworkPolicy as a Kubernetes API might be expected to be. I think the NodePort example really gets to the crux of this issue. We have previously talked about this as whatever source IP the pod itself sees - that's useful for implementations, and not _entirely_ broken for users, but a more usable API would maybe extend a little bit further - i.e., the source IP as it enters the Kubernetes cluster. It's probably not fair to say NetworkPolicy should be aware of devices outside of the Kubernetes cluster itself, though.

Reply:
+1 to what Casey said here. While it is not possible to enforce the policy on original source IP if the SNAT happens outside the cluster (ex: cloud-provider Load Balancer), we can certainly define the API to consider the source IP as it enters kubernetes cluster.

Considering a specific traffic flow for node port,`EXTERNAL-IP -> NODE-IP1:NODE-PORT` that gets rewritten to NODE-IP1 -> POD2(on NODE2):PORT. In this case, the source IP is considered to be NODE-IP1 whereas the actual IP address is an external IP address. This not only allows all external IP addresses that are allowed by cloud-provider firewall but also interfere with node based policies that we introduce in future.

Reply:
For completeness, the reason we haven't been able to take such a strong stance to-date is because network policy works in terms of _pods_ and not services. 

Take for example a service that selects both pod A and pod B, but each of those pods has different network policy applied. Most implementations will need to understand which of those pods will be the ultimate destination in order to apply the correct policy. At the point that traffic enters the cluster, that may not be knowable (e.g., since kube-proxy will arbitrarily assign the traffic to one of those pods).

Reply:
Yeah I agree that it will introduce additional complexities for CNI implementers.

I think the enforcement can happen at two places,
1. The node that first received the packet
2. The node that hosts the backend pod

The service resolution happens on the node that first received the packet. If the policy enforcement is to happen after the service resolution on this node, then we know about the backend pod.

If the enforcement happen on the node that hosts the backend pod, the CNI implementer need to use global conntrack state or some other mechanism to identify the original source IP.

In both cases, there seems to be a way to know about the backend pod depending on where in the stack the policy is enforced.

Reply:
@abhiraut so I want to make it clear that the above was more of a "this is probably what is best" type discussion.

I believe we have already defined the `ipBlock` semantics for the existing NetworkPolicy API to be "as seen by the selected pod" - although I know this was discussed when originally implementing the API, this doesn't seem to be written down anywhere, and so I can't prove it. .  Are you suggesting with this KEP that we change the semantic for both? Or that we have different behaviors for CNP vs regular NP?

I think the former would be great, if we can do it without breaking API compatibility contracts. I think the latter would be a bad outcome.

More generally though, I think re-deifning what `ipBlock` means is arguably out of scope for this KEP.

Reply:
> I believe we have already defined the `ipBlock` semantics for the existing NetworkPolicy API to be "as seen by the selected pod" - although I know this was discussed when originally implementing the API, this doesn't seem to be written down anywhere, and so I can't prove it.

It's in https://kubernetes.io/docs/concepts/services-networking/network-policies/#behavior-of-to-and-from-selectors.

Reply:
> I'm not at all convinced that "as it enters the Kubernetes cluster" is realistic.

Possibly, but if so, then this KEP does not solve the problems of the users in its User Stories, and so should be rejected (or at least, rewritten to not claim to be solving those problems, in which case parts of it may no longer make sense as part of a solution that is only trying to solve the remaining user stories, etc).

> something about trusting the various ingress mechanisms to all be covered makes me uneasy.

When I was trying to write up a KEP for egress-only admin-focused network policy a while back ([ClusterEgressFirewall](https://github.com/danwinship/enhancements/blob/cluster-egress-firewall/keps/sig-network/20190917-cluster-egress-firewall.md)) I said:

> Unlike with NetworkPolicy, which is expected to be implemented entirely by the network plugin, ClusterEgressFirewall will be a shared responsibility; any component of Kubernetes that provides a way for packets to exit the cluster (eg, secondary network interfaces) must ensure that egress firewall rules will be obeyed by exiting traffic. Any component that has the effect of obscuring the origin of packets (eg, proxies or load balancers) must ensure that they cannot be used to bypass egress firewall rules.

(The "secondary network interfaces" part is probably definitely out of scope for ClusterNetworkPolicy but the various issues around using cloud load balancers to bypass policy rules still apply.)

Reply:
if I have a proxyful LB implementation, and I set up an ipBlock rule, the proxy IP must be included in the rule, but proxy IPs are not a useful unit to block.  Saying that the proxyful LB must respect NetworkPolicy is maybe appropriate but is a huge net-new requirement and we'd have to think really hard about whether it was feasible or not.

Or we simply don't - if you are exposing something "externally" then there's a different expression of policy needed (e.g. pod selector does not make sense)